courseworks/week4/shiqinruixu/week4.ipynb

```diff
@@ -0,0 +1,182 @@
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "text/plain": [
+       "5×5 Matrix{Float64}:\n",
+       " 0.502475  -0.150831      0.0236101   0.0           0.0\n",
+       " 0.0        1.15588      -0.341383    0.195486      0.0\n",
+       " 0.0        2.77556e-17   0.949328   -2.27544      -0.00647398\n",
+       " 0.0        0.0           0.0         0.0212472    -0.17786\n",
+       " 0.0        0.0           0.0         8.67362e-19  -0.504495"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[0m\u001b[1mTest Summary: | \u001b[22m\u001b[32m\u001b[1mPass  \u001b[22m\u001b[39m\u001b[36m\u001b[1mTotal  \u001b[22m\u001b[39m\u001b[0m\u001b[1mTime\u001b[22m\n",
+      "test givens   | \u001b[32m   2  \u001b[39m\u001b[36m    2  \u001b[39m\u001b[0m1.3s\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "Test.DefaultTestSet(\"test givens\", Any[], 2, false, false, true, 1.683561520903971e9, 1.683561522183982e9)"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "using LinearAlgebra,Test, Random\n",
+    "\n",
+    "struct GivensMatrix{T} <: AbstractArray{T, 2}\n",
+    "\tc::T\n",
+    "\ts::T\n",
+    "\ti::Int\n",
+    "\tj::Int\n",
+    "\tn::Int\n",
+    "end\n",
+    "\n",
+    "Base.size(g::GivensMatrix) = (g.n, g.n)\n",
+    "\n",
+    "Base.size(g::GivensMatrix, i::Int) = i == 1 || i == 2 ? g.n : 1\n",
+    "\n",
+    "function givens(A, i, j)\n",
+    "\tx, y = A[i, 1], A[j, 1]\n",
+    "\tnorm = sqrt(x^2 + y^2)\n",
+    "\tc = x/norm\n",
+    "\ts = y/norm\n",
+    "\treturn GivensMatrix(c, s, i, j, size(A, 1))\n",
+    "end\n",
+    "\n",
+    "function left_mul!(A::AbstractMatrix, givens::GivensMatrix)\n",
+    "\t@inbounds for col in 1:min(size(A, 2),3)\n",
+    "\t\tvi, vj = A[givens.i, col], A[givens.j, col]\n",
+    "\t\tA[givens.i, col] = vi * givens.c + vj * givens.s\n",
+    "\t\tA[givens.j, col] = -vi * givens.s + vj * givens.c\n",
+    "\tend\n",
+    "\treturn A\n",
+    "end\n",
+    "function right_mul!(A::AbstractMatrix, givens::GivensMatrix)\n",
+    "\t@inbounds for row in 1:size(A, 1)\n",
+    "\t\tvi, vj = A[row, givens.i], A[row, givens.j]\n",
+    "\t\tA[row, givens.i] = vi * givens.c + vj * givens.s\n",
+    "\t\tA[row, givens.j] = -vi * givens.s + vj * givens.c\n",
+    "\tend\n",
+    "\treturn A\n",
+    "end\n",
+    "\n",
+    "function givens_trisym!(Q::AbstractMatrix,A::AbstractMatrix)\n",
+    "    m, n = size(A)\n",
+    "\tif m == 1\n",
+    "\t\treturn Q, A\n",
+    "\telse\n",
+    "\t\tg = givens(A, 1, 2)\n",
+    "\t\tleft_mul!(A, g)\n",
+    "\t\tright_mul!(Q, g)\n",
+    "\n",
+    "\t\tgivens_trisym!(view(Q, :, 2:m), view(A, 2:m, 2:n))\n",
+    "\t\treturn Q, A\n",
+    "\tend\n",
+    "end\n",
+    "\n",
+    "@testset \"test givens\" begin\n",
+    "    n = 5\n",
```

Comment:
Well done!
A suggestion: when testing your code you should choose more random cases, for example:
```
@testset "Givens rotation" begin
    for i in 1:10
        n = rand(1:100) # this is the dimension of your matrix
        @test test_func(n) # this is your test function
    end
end
```